### DISCUSSION TITLE
Interaction with git-imerge

### DISCUSSION THREAD

**Discussion by pokey** [arxanas/git-branchless#380] [Q&A] [answered]
When there is a merge conflict, I prefer to use [git-imerge](https://github.com/mhagger/git-imerge) to resolve it:

```sh
git switch <branch>
git-imerge rebase main
```

However, this workflow leaves a bunch of garbage commits from the imerge process in my smartlog.  What's the best way to hide them all at once?

Thanks!

**Comment by claytonrcarter** [comment]
You can look into using `git hide -r <first garbage commit on old branch>`.

Short of that, I've been tinkering w/ some other flags locally to make it easier to bulk hide commits, but I haven't stabilized them yet.

**Reply by pokey** [reply]
Yeah I looked into hide -r, but I don't think that solves the problem here, because I don't believe there's a common commit that will be a parent of all the garbage commits but not the ones that I want

**Reply by pokey** [reply]
See for example:

<img width="815" alt="image" src="https://user-images.githubusercontent.com/755842/165540594-8ec539e0-7259-4357-8320-205d51b8feab.png">

Note how there are garbage commits hanging off of every commit on the `main` branch

**Reply by claytonrcarter** [reply]
Oh, that's juicy. So `imerge` is generating both the `imerge` and `Merge commit` commits, none of which you want? And are you working w/ branches for most of this, or building commits w/o a branch pointer?

If you're working w/ branches, one of the flags I've been playing w/ is `--prune`, which is intended to hide all commits that aren't "on a branch", like `git hide --prune @` to hide any dangling/non-branch commits on my current branch. 

As is, though, your best bet is to either do the log grep trick that you mentioned, or use `git hide -r` but then manually copy/paste each of the root commits for the garbage trees. From this example, `git hide -r 60af92e 3a191e3 5d3c326` would (I believe) end up hiding all of the `imerge`/`Merge commit` commits.

**Reply by pokey** [reply]
Ah prune is an interesting idea. Seems slightly counter to the "branchless" name tho 😅. But I'm one to talk: I work on something called Cursorless, which just so happens to be at its most powerful when used with multiple cursors 😄

But yeah prune could be useful for my other complaint, which is that when I merge a branch via GitHub squash, I have to both delete local branch and then hide the commits. Which obviously wouldn't be a thing if I were properly branchless, but 🤷

**Comment by pokey** [comment]
I wonder if there's some way to temporarily disable tracking?  Could maybe use that while doing the imerge?

**Reply by claytonrcarter** [reply]
Do you know if `imerge` calls git hooks (like `post-rewrite`, etc)? Hooks like that are how `branchless` is able to keep track of which commits are rewritten, discarded, etc. If a 3rd party tool isn't calling hooks, then it's difficult (impossible?) for `branchless` to know that something has happened. I believe that `git revise` has a similar issue: https://github.com/mystor/git-revise/issues/35

**Reply by pokey** [reply]
I would assume that it does. I see lots of git branchless spam during the imerge process

**Reply by arxanas** [reply]
@pokey Unfortunately, there's currently no way to skip commit detection in git-branchless, but it should be fairly simple to implement, if you're interested. It should just be a matter of checking to see if some environment variable `BRANCHLESS_XXX` is set at the beginning of this function and then bailing early if it is:

https://github.com/arxanas/git-branchless/blob/c4fb1a476d80c2fea7e0e37149b8a8efb7f0b322/git-branchless/src/commands/hooks.rs#L69

**Comment by pokey** [comment]
This isn't pretty, but seems to do the trick if the imerge was in the past 30 minutes and you are rebasing (so there are no merge commits in your desired output history):

```bash
git hide $(git log --all --since="30 minutes ago" '--grep=^Merge' '--grep=^imerge' --format=format:%H)
```

**Comment by pokey** [comment, marked as answer]
Just an update in case someone lands here: as mentioned in https://github.com/arxanas/git-branchless/discussions/496, this command should work:

```
$ git hide 'descendants(draft() & message("regex:^Merge commit|^imerge") - ancestors(branches()))'
```

**Reply by arxanas** [reply]
Neat!
